util/src/openssl_utils.cpp

```diff
@@ -0,0 +1,258 @@
+
+  // Get the local stored certificate for this peer
+  string local_cert_path =
+      (use_unified_certs)
+          ? cert_root_directory + "/" + std::to_string(remotePeerId) + "/" + "node.cert"
+          : cert_root_directory + "/" + std::to_string(remotePeerId) + "/" + cert_type + "/" + cert_type + ".cert";
+  auto deleter = [](FILE* fp) {
+    if (fp) fclose(fp);
+  };
+  std::unique_ptr<FILE, decltype(deleter)> fp(fopen(local_cert_path.c_str(), "r"), deleter);
+  if (!fp) {
+    LOG_ERROR(OPENSSL_LOG, "Certificate file not found, path: " << local_cert_path);
+    return false;
+  }
+
+  UniqueOpenSSLX509 localCert(PEM_read_X509(fp.get(), NULL, NULL, NULL));
+  if (!localCert) {
+    LOG_ERROR(OPENSSL_LOG, "Cannot parse certificate, path: " << local_cert_path);
+    return false;
+  }
+
+  // this is actual comparison, compares hash of 2 certs
+  bool res = (X509_cmp(cert_to_verify, localCert.get()) == 0);
+  return res;
+}
+
+pair<string, string> OpenSSLCryptoImpl::generateEdDSAKeyPair(const KeyFormat fmt) const {
+  UniquePKEY edPkey;
+  UniqueOpenSSLPKEYContext edPkeyCtx(EVP_PKEY_CTX_new_id(NID_ED25519, nullptr));
+
+  ConcordAssertNE(edPkeyCtx, nullptr);
+
+  ConcordAssertEQ(1, EVP_PKEY_keygen_init(edPkeyCtx.get()));
+  ConcordAssertEQ(
+      1, EVP_PKEY_keygen(edPkeyCtx.get(), reinterpret_cast<EVP_PKEY**>(&edPkey)));  // Generate EdDSA key 'edPkey'.
+
+  unsigned char privKey[EdDSASignatureByteSize]{};
+  unsigned char pubKey[EdDSASignatureByteSize]{};
+  size_t privlen{EdDSASignatureByteSize};
+  size_t publen{EdDSASignatureByteSize};
+
+  ConcordAssertEQ(1, EVP_PKEY_get_raw_private_key(edPkey.get(), privKey, &privlen));
+  ConcordAssertEQ(1, EVP_PKEY_get_raw_public_key(edPkey.get(), pubKey, &publen));
+
+  pair<string, string> keyPair;
+  keyPair.first = concordUtils::bufferToHex(reinterpret_cast<const char*>(privKey), (const size_t)privlen, false);
+  keyPair.second = concordUtils::bufferToHex(reinterpret_cast<const char*>(pubKey), (const size_t)publen, false);
+
+  if (KeyFormat::PemFormat == fmt) {
+    keyPair = EdDSAHexToPem(keyPair);
+  }
+  return keyPair;
+}
+
+pair<string, string> OpenSSLCryptoImpl::EdDSAHexToPem(const std::pair<std::string, std::string>& hex_key_pair) const {
+  string privPemString;
+  string pubPemString;
+
+  if (!hex_key_pair.first.empty()) {  // Proceed with private key pem file generation.
+    const auto privKey = concordUtils::unhex(hex_key_pair.first);
+
+    UniquePKEY ed_privKey(EVP_PKEY_new_raw_private_key(NID_ED25519, nullptr, privKey.data(), privKey.size()));
+
+    ConcordAssertNE(nullptr, ed_privKey);
+
+    const char* const tempPrivPem = "/tmp/concordTempPrivKey.pem";
+
+    FILE* fp = fopen(tempPrivPem, "w");
+    PEM_write_PrivateKey(fp, ed_privKey.get(), nullptr, nullptr, 0, nullptr, nullptr);
+    fclose(fp);
+
+    // Read private key from pem file.
+    std::ifstream privPem(tempPrivPem);
+    if (privPem.is_open()) {
+      string temp;
+
+      while (privPem.good()) {
```

Comment:
Assuming you want to read the whole file, why not do something like:
std::ifstream t("file");
std::stringstream buffer;
buffer << t.rdbuf(); 
?

Do note that reading a file like this is unsafe (what if for some reason it contains a lot of data?)
you should limit the number of bytes you read.

Reply:
Code changed to use `std::string readFile(FILE* file, uint64_t maxBytes);`.